src/validation.cpp

```diff
@@ -3129,19 +3138,23 @@ void CChainState::ProcessLoanEvents(const CBlockIndex* pindex, CCustomCSView& ca
             }
 
             // All done. Ready to save the overall auction.
-            cache.StoreAuction(vaultId, CAuctionData{
-                                            uint32_t(batches.size()),
-                                            pindex->nHeight + chainparams.GetConsensus().blocksCollateralAuction(),
-                                            cache.GetLoanLiquidationPenalty()
-            });
+            cache.StoreAuction(vaultId, {uint32_t(batches.size()), pindex->nHeight + auctionDuration, cache.GetLoanLiquidationPenalty()});
 
             // Store state in vault DB
-            if (pvaultHistoryDB) {
-                pvaultHistoryDB->WriteVaultState(cache, *pindex, vaultId, collateral.val->ratio());
+            if (vaultHistory) {
+                vaultHistory->WriteVaultState(cache, *pindex, vaultId, collateral.val->ratio(), batches);
             }
+        };
 
+        cache.ForEachVaultCollateral([&](const CVaultId& vaultId, const CBalances& collaterals) {
```

Comment:
Is it possible to just shard the vaults by the number of hardware concurrency, and provide it a prefix and end key to each before hand, and just let the thread run, and then join them all in the end? 

That would be a faster implementation, since you don't have so much message passing between each threads. It's just one set in the beginning and end.